man/systemd.resource-control.xml

```diff
@@ -710,6 +716,42 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>BPFProgram=<replaceable>BPF_ATTACH_TYPE</replaceable><constant>:</constant><replaceable>BPF_FS_PROGRAM_PATH</replaceable></varname></term>
+        <listitem>
+          <para>Add a custom cgroup BPF program.</para>
+
+          <para><varname>BPFProgram=</varname> allows attaching cgroup-bpf hooks to a systemd unit including
+          egress and ingress attachment points also exposed to <varname>IP(Egress|Ingress)FilterPath=</varname>.
+          Cgroup-bpf hooks in the form of BPF programs loaded to the BPF filesystem are attached with cgroup-bpf attach
+          flags determined by the unit. For details about attachment types and flags see <ulink
+          url="https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/plain/include/uapi/linux/bpf.h"/></para>
+
+          <para>The specification of BPF program consists of a <replaceable>BPF_ATTACH_TYPE</replaceable> followed by a
+          <replaceable>BPF_FS_PROGRAM_PATH</replaceable> with <literal>:</literal> as the separator:
+          <replaceable>BPF_ATTACH_TYPE</replaceable><constant>:</constant><replaceable>BPF_FS_PROGRAM_PATH</replaceable>.</para>
+
+          <para><varname>BPF_ATTACH_TYPE</varname> is the accustomed string name of BPF attach type also used in
+          <command>bpftool</command>. <varname>BPF_ATTACH_TYPE</varname> can be one of <constant>egress</constant>,
+          <constant>ingress</constant>, <constant>sock_create</constant>, <constant>sock_ops</constant>,
+          <constant>device</constant>, <constant>bind(4|6)</constant>, <constant>connect(4|6)</constant>,
+          <constant>post_bind(4|6)</constant>, <constant>sendmsg(4|6)</constant>, <constant>sysctl</constant>,
+          <constant>recvmsg(4|6)</constant>, <constant>(get|set)sockopt</constant>.</para>
+
+          <para>Setting <varname>BPFProgram=</varname> to an empty value makes previous assignments ineffective.</para>
+          <para>If the same <replaceable>BPF_FS_PROGRAM_PATH</replaceable>:<replaceable>BPF_ATTACH_TYPE</replaceable>
+          value is set multiple times, only first assignment will be effective and only one program will be attached to
+          cgroup hook. If BPF <constant>egress</constant> program pinned to <replaceable>BPF_FS_PROGRAM_PATH</replaceable>
```

Comment:
@keszybz 

Only one assignment will be effective, the one passed first:

https://github.com/systemd/systemd/pull/13496/commits/329a5dd70791a081db38be6d1a05c833151ef580#diff-11d3c4581343e0c3d561b27621161777ac95427feb2dd615b3c2f59429e80aefR100-R104

Reply:
Let's look at this from users POV (which is what the docs describe):
```
BPFProgram=egress:/sys/fs/bpf/egress-hook
BPFProgram=egress:/sys/fs/bpf/egress-hook
```

This paragraph now says that the first line is effective. Technically, this is what happens because of the implementation. But from user's POV, this is not distinguishable from the second line being effective.

Reply:
@keszybz 

I think we could say that multiple assignments of the same program are idempotent, but that sounds scary, hence:

> Multiple assignments of the same `TYPE`:`PROGRAM_PATH` value have the same effect as a single assignment: the program with the path `PROGRAM_PATH` will be atttached to cgroup hook `TYPE` just once.